compiler/module_graph.rs

```diff
@@ -0,0 +1,1022 @@
+  fn fetch(&mut self, specifier: &ModuleSpecifier) -> FetchFuture;
+
+  /// Get a module from cache, which has been previously fetched. It is expected
+  /// that source for the module had been invalidated that the related emitted
+  /// code would not be returned either.
+  fn get_cache(&self, specifier: &ModuleSpecifier) -> Result<CachedModule>;
+
+  /// Get the optional build info from the cache for a given module specifier.
+  /// Because build infos are only associated with the "root" modules, they are
+  /// not expected to be cached for each module, but are "lazily" checked when
+  /// a root module is identified.  The `cache_type` also indicates what form
+  /// of the module the build info is valid for.
+  fn get_build_info(
+    &self,
+    specifier: &ModuleSpecifier,
+    cache_type: &CacheType,
+  ) -> Result<Option<String>>;
+
+  /// Set the emitted code (and maybe map) for a given module specifier.  The
+  /// cache type indicates what form the emit is related to.
+  fn set_cache(
+    &mut self,
+    specifier: &ModuleSpecifier,
+    cache_type: &CacheType,
+    code: String,
+    maybe_map: Option<String>,
+  ) -> Result<()>;
+
+  /// When parsed out of a JavaScript module source, the triple slash reference
+  /// to the types should be stored in the cache.
+  fn set_types(
+    &mut self,
+    specifier: &ModuleSpecifier,
+    types: String,
+  ) -> Result<()>;
+
+  /// Set the build info for a module specifier, also providing the cache type.
+  fn set_build_info(
+    &mut self,
+    specifier: &ModuleSpecifier,
+    cache_type: &CacheType,
+    build_info: String,
+  ) -> Result<()>;
+
+  /// Set the graph dependencies for a given module specifier.
+  fn set_deps(
+    &mut self,
+    specifier: &ModuleSpecifier,
+    dependencies: CachedDependencies,
+  ) -> Result<()>;
+}
+
+impl std::fmt::Debug for dyn SpecifierHandler {
+  fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+    write!(f, "SpecifierHandler {{ }}")
+  }
+}
+
+lazy_static! {
+  /// Matched the `@deno-types` pragma.
+  static ref DENO_TYPES_RE: Regex =
+    Regex::new(r#"(?i)^\s*@deno-types\s*=\s*(?:["']([^"']+)["']|(\S+))"#)
+      .unwrap();
+  /// Matches a `/// <reference ... />` comment reference.
+  static ref TRIPLE_SLASH_REFERENCE_RE: Regex =
+    Regex::new(r"(?i)^/\s*<reference\s.*?/>").unwrap();
+  /// Matches a path reference, which adds a dependency to a module
+  static ref PATH_REFERENCE_RE: Regex =
+    Regex::new(r#"(?i)\spath\s*=\s*["']([^"']*)["']"#).unwrap();
+  /// Matches a types reference, which for JavaScript files indicates the
+  /// location of types to use when type checking a program that includes it as
+  /// a dependency.
+  static ref TYPES_REFERENCE_RE: Regex =
+    Regex::new(r#"(?i)\stypes\s*=\s*["']([^"']*)["']"#).unwrap();
+}
+
+#[derive(Clone, Debug)]
+pub struct CachedDependency {
+  pub specifier: String,
+  pub maybe_code: Option<ModuleSpecifier>,
+  pub maybe_type: Option<ModuleSpecifier>,
```

Comment:
What does this field describe? Is it `@deno-types` directive's value?

Reply:
Yes, or a `.d.ts` import, as they both go into slots that shouldn't have runtime code, but type only code.

Reply:
Actually, I lied... it isn't `.d.ts` files, it is any `import type`/`export type` go into that slot in addition to `@deno-types`.